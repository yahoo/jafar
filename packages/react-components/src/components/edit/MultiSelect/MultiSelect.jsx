/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import isEqual from 'lodash/isEqual';
import noop from 'lodash/noop';
import _Select from 'react-select';
import { customStyles } from '../Select/Select';
import { toJafar } from '../../utils';
 
const propTypes = {
  value: PropTypes.arrayOf(PropTypes.any),
  state: PropTypes.shape({
    items: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.string,
      value: PropTypes.any,
    })),
    placeholder: PropTypes.string,
    searchable: PropTypes.bool,
    searchQuery: PropTypes.string,
    styles: PropTypes.object,
  }),
  disabled: PropTypes.bool,
  onValueChange: PropTypes.func.isRequired,
  onStateChange: PropTypes.func,
};

const defaultProps = {
  value: [],
  state: {
    items: [],
    placeholder: 'Search',
    searchable: false,
    searchQuery: '',
    styles: {},
  },
  disabled: false,
  onStateChange: noop,
};
 
export const mapper = ({ value, state, disabled, onValueChange, onStateChange }) => {
  const onChange = selected => {
    const newValue = !selected ? undefined : selected.map(item => item.orgValue);
    onValueChange(newValue);
  };

  const onInputChange = searchQuery => onStateChange(({ state }) => ({ ...state, searchQuery }));

  // for value that is array of object we use itemIdField, cuz internal select should get value as string - 
  // it has problem getting objects as value since it cant compare between them
  const items = (state.items || []).map((item) => {
    return {
      label: item.label, 
      value: state.itemIdField ? item.value[state.itemIdField] : item.value, 
      orgValue: item.value, 
    };
  });

  let selected = (value || []).map(selectedItemValue => 
    items.find(item => isEqual(selectedItemValue, item.orgValue)));

  let options = items.filter(item => 
    !selected.find(selectedItem => isEqual(selectedItem.value, item.orgValue)));

  const styles = customStyles(state.styles);

  return {
    isMulti: true,
    value: selected, 
    options: options,
    placeholder: state.placeholder,
    isDisabled: disabled,
    isSearchable: !!state.searchable,
    inputValue: state.searchQuery,
    styles: styles,
    onChange,
    onInputChange,
  };
};
 
const MultiSelect = toJafar(_Select, mapper);
 
MultiSelect.propTypes = propTypes;
MultiSelect.defaultProps = defaultProps;
 
export default MultiSelect;
 
// hack for styleguidist (issue with hoc components)
/** @component */
/**
  * Represent a selection of multi items from async items list 
  * 
  * Import <a target="_blank" 
  href="https://github.com/yahoo/jafar/blob/master/packages/react-components/src/components/edit/MultiSelect">
  MultiSelect</a> from '@jafar/react-components/edit/MultiSelect'
  */
export const DemoMultiSelect = props => <MultiSelect {...props} />;
DemoMultiSelect.propTypes = propTypes;
DemoMultiSelect.defaultProps = defaultProps;
DemoMultiSelect.displayName = 'MultiSelect';
 
