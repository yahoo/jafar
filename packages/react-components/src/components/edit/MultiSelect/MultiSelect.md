<h3>Examples</h3>

<h4>Usage in jafar form</h4>

```javascript
const Form = require('@jafar/react-form/Form').default;
const Field = require('@jafar/react-form/Field').default;

const model = {
  id: 'simple',
  fields: {
    bestFriends: {
      label: 'Best Friends',
      path: 'bestFriends',
      component: {
        name: 'MultiSelect',
        state: {
        items: [{
                value: '1',
                label: 'Ross Geller'
            }, {
                value: '2',
                label: 'Monica Geller'
            }, {
                value: '3',
                label: 'Rachel Green'
            }, {
                value: '4',
                label: 'Chandler Bing'
            }, {
                value: '5',
                label: 'Joey Tribbiani'
            }, {
                value: '6',
                label: 'Phoebe Buffay'
            }],
        },
      },
      required: true,
      disableTerm: {
          name: 'disableMe',
      },
      excludeTerm: {
          name: 'excludeMe',
      },
    },
  },
  data: {
    bestFriends: ['2', '3'], // remove to see initial value undefined, and required error
  },
};

const resources = {
  components: { 
    MultiSelect: {
      renderer: MultiSelect,
    },
  },
  terms: {
      disableMe: { func: () => false }, // convert to true to see the field disabled
      excludeMe: { func: () => false }, // convert to true to see the field excluded
  },
};

<Form model={model} resources={resources}>
    <Field id='bestFriends' />
</Form>
```

<h4>Simple - fixed items</h4>

```javascript
initialState = { 
  value: ['2', '3'],
  state: {
    items:[{
        value: '1',
        label: 'Ross Geller'
    }, {
        value: '2',
        label: 'Monica Geller'
    }, {
        value: '3',
        label: 'Rachel Green'
    }, {
        value: '4',
        label: 'Chandler Bing'
    }, {
        value: '5',
        label: 'Joey Tribbiani'
    }, {
        value: '6',
        label: 'Phoebe Buffay'
    }],
  },
};

<MultiSelect
value={state.value}
state={state.state}
onValueChange={(value) => {
  setState({ value });
}}
/>

```

<h4>Value as object</h4>

```javascript
initialState = { 
  value: [{ id: '3', first: 'Rachel', last: 'Green' }],
  state: {
    itemIdField: 'id',
    items:[{
        value: { id: '1', first: 'Ross', last: 'Geller' },
        label: 'Ross Geller'
    }, {
        value: { id: '2', first: 'Monica', last: 'Geller' },
        label: 'Monica Geller'
    }, {
        value: { id: '3', first: 'Rachel', last: 'Green' },
        label: 'Rachel Green'
    }, {
        value: { id: '4', first: 'Chandler', last: 'Bing' },
        label: 'Chandler Bing'
    }, {
        value: { id: '5', first: 'Joey', last: 'Tribbiani' },
        label: 'Joey Tribbiani'
    }, {
        value: { id: '6', first: 'Phoebe', last: 'Buffay' },
        label: 'Phoebe Buffay'
    }],
  },
};

<MultiSelect
value={state.value}
state={state.state}
onValueChange={(value) => {
  setState({ value });
}}
/>
```

<h4>Searchable</h4>

```javascript
initialState = { 
  value: ['3'],
  state: {
    searchable: true,
    items:[{
        value: '1',
        label: 'Ross Geller'
    }, {
        value: '2',
        label: 'Monica Geller'
    }, {
        value: '3',
        label: 'Rachel Green'
    }, {
        value: '4',
        label: 'Chandler Bing'
    }, {
        value: '5',
        label: 'Joey Tribbiani'
    }, {
        value: '6',
        label: 'Phoebe Buffay'
    }],
  },
};

<MultiSelect
value={state.value}
state={state.state}
onValueChange={(value) => {
  setState({ value });
}}
/>
```

<h4>Disabled</h4>

```javascript
initialState = { 
  value: ['3', '4'],
  state: {
    items:[{
        value: '1',
        label: 'Ross Geller',
    }, {
        value: '2',
        label: 'Monica Geller'
    }, {
        value: '3',
        label: 'Rachel Green'
    }, {
        value: '4',
        label: 'Chandler Bing'
    }, {
        value: '5',
        label: 'Joey Tribbiani'
    }, {
        value: '6',
        label: 'Phoebe Buffay'
    }],
  },
};

<MultiSelect
value={state.value}
state={state.state}
disabled={true}
onValueChange={(value) => {
  setState({ value });
}}
/>
```

  <h4>Value undefined</h4>

```javascript
initialState = { 
  value: undefined,
  state: {
    items:[{
        value: '1',
        label: 'Ross Geller'
    }, {
        value: '2',
        label: 'Monica Geller'
    }, {
        value: '3',
        label: 'Rachel Green'
    }, {
        value: '4',
        label: 'Chandler Bing'
    }, {
        value: '5',
        label: 'Joey Tribbiani'
    }, {
        value: '6',
        label: 'Phoebe Buffay'
    }],
  },
};

<MultiSelect
value={state.value}
state={state.state}
onValueChange={(value) => {
  setState({ value });
}}
/>
```