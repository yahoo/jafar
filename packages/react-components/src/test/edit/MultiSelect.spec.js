/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { shallow } from 'enzyme';
import MultiSelect from '../../components/edit/MultiSelect';
import { mapper } from '../../components/edit/MultiSelect/MultiSelect';
  
describe('MultiSelect', () => {
  let componentProps;
  
  const jafarProps = {
    value: ['ocean'],
    disabled: false,
    required: false,
    state: {
      searchable: true,
      searchQuery: 'oc',
      placeholder: 'Color...',
      items:[
        { value: 'ocean', label: 'Ocean' },
        { value: 'blue', label: 'Blue' },
        { value: 'purple', label: 'Purple' },
      ],
    },
    onValueChange: jest.fn(),
    onStateChange: jest.fn(),
  };

  const mapFunc = (item) => ({ label: item.label, value: item.value, orgValue: item.value });
  
  const expectedProps = {
    isMulti: true,
    value: [{ value: 'ocean', label: 'Ocean' }].map(mapFunc),
    options: jafarProps.state.items.slice(1).map(mapFunc),
    placeholder: 'Color...',
    isDisabled: false,
    isSearchable: true,
    inputValue: 'oc',
    styles: expect.any(Object),
    onChange: expect.any(Function),
    onInputChange: expect.any(Function),
  };
   
  beforeEach(() => {
    componentProps = mapper(jafarProps);
  });
  
  describe('mapper', () => {
    it('return correct props', () => {
      expect(componentProps).toEqual(expectedProps);
    });
    
    it('onChange - call onValueChange with correct value', () => {
      const selected = expectedProps.options[0];
      componentProps.onChange([selected]);
      expect(jafarProps.onValueChange).toHaveBeenCalledWith([jafarProps.state.items[1].value]);
    });
 
    it('onChange - call onValueChange with undefined value', () => {
      componentProps.onChange(undefined);
      expect(jafarProps.onValueChange).toHaveBeenCalledWith(undefined);
    });
 
    it('onInputChange - call onStateChange with correct value', () => {
      componentProps.onInputChange('ra');
      expect(jafarProps.onStateChange).toHaveBeenCalledWith(expect.any(Function));
      const updater = jafarProps.onStateChange.mock.calls[0][0];
      expect(updater({ state: { items: [] } })).toEqual({ items: [], searchQuery: 'ra' });
    });
  });
  
  describe('component', () => {
    it('renders ok', () => {
      const component = shallow(<MultiSelect {...jafarProps} />);
      expect(component.props()).toMatchObject(expectedProps);
    });
  });
});
