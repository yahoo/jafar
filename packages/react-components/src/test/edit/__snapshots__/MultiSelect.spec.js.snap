// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultiSelect /> Should render provided data - disabled 1`] = `
<StateManager
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  isDisabled={true}
  isMulti={true}
  isSearchable={false}
  onChange={[Function]}
  onInputChange={[Function]}
  options={Array []}
  styles={
    Object {
      "clearIndicator": [Function],
      "control": [Function],
      "dropdownIndicator": [Function],
      "menu": [Function],
      "menuList": [Function],
      "singleValue": [Function],
      "valueContainer": [Function],
    }
  }
  value={
    Array [
      undefined,
    ]
  }
/>
`;

exports[`<MultiSelect /> Should render provided data - required 1`] = `
<StateManager
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  isDisabled={false}
  isMulti={true}
  isSearchable={false}
  onChange={[Function]}
  onInputChange={[Function]}
  options={Array []}
  styles={
    Object {
      "clearIndicator": [Function],
      "control": [Function],
      "dropdownIndicator": [Function],
      "menu": [Function],
      "menuList": [Function],
      "singleValue": [Function],
      "valueContainer": [Function],
    }
  }
  value={
    Array [
      undefined,
    ]
  }
/>
`;

exports[`<MultiSelect /> Should render provided data 1`] = `
<StateManager
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  isDisabled={false}
  isMulti={true}
  isSearchable={false}
  onChange={[Function]}
  onInputChange={[Function]}
  options={Array []}
  styles={
    Object {
      "clearIndicator": [Function],
      "control": [Function],
      "dropdownIndicator": [Function],
      "menu": [Function],
      "menuList": [Function],
      "singleValue": [Function],
      "valueContainer": [Function],
    }
  }
  value={
    Array [
      undefined,
    ]
  }
/>
`;
